design/pushed-authorization-requests.md

```diff
@@ -0,0 +1,326 @@
+* org.keycloak.models.jpa.entities.ClientEntity
+* org.keycloak.models.map.client.AbstractClientEntity
+
+### The PAR endpoint 
+
+(1) : Authenticate the client in the same way as at the authorization endpoint
+
+(2) : Accept The OAuth 2.0 authorization request parameters
+
+(3) : Reject the request if the "request_uri" authorization request parameter is provided.
+
+(4) : Validate the pushed request as it would an authorization request sent to the authorization endpoint
+
+(5) : Generate request_uri
+
+(6) : save the Auth Request+request_uri+request_uri_lifespan
+
+(7) : return PAR Response
+
+Classes/methods to be added:
+
+```java
+/**
+ * OAuth 2.0 PAR request endpoint
+ */
+public class ParEndpoint extends AuthorizationEndpointBase {
+
+    /**
+     * Handles PAR requests.
+     *
+     * @return the PAR response.
+     */
+    @Path("par")
+    @POST
+    @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response handleParRequest() {
+        
+    }
+    
+}
+```
+
+### Generation of request_uri
+As per spec,
+
+The format of the "request_uri" value is at the discretion of the authorization server but it MUST contain some part generated using a cryptographically strong pseudorandom algorithm such that it is computationally infeasible to predict or guess a valid value. The authorization server MAY construct the "request_uri" value using the form "urn:ietf:params:oauth:request_uri:<reference-value>" with "<reference-value>" as the random part of the URI that references the respective authorization request data. The string representation of a UUID as a URN per [RFC4122] is also an option for authorization servers to construct "request_uri" values. The "request_uri" value MUST be bound to the client that posted the authorization request.
+
+Classes to be added:
+ * org.keycloak.common.util.RequestUriUtil
+
+### Save the authorization request with the associated request_uri generated + request_uri_lifespan
+It can be done with org.keycloak.sessions.AuthenticationSessionModel
+
+```
+
+authenticationSession.setClientNote(request_uri, request+request_uri_lifespan);
+
+```
+
+### Retrieve the authorization request with the request_uri
+
+At the Authorization Endpoint, authorization server must be able to retrieve the authorization request with the request_uri.
+It can be done with org.keycloak.sessions.AuthenticationSessionModel
+
+```
+
+authenticationSession.getClientNote(request_uri);
+
+```
+
+
+### Change in Authorization endpoint
+Authorisation server should also check request_uri combine with:
+
+* Server metadata require_pushed_authorization_requests 
+* Client metadata require_pushed_authorization_requests
+* request_uri_lifespan
+
+Authorisation server should also be able to retrieve authorisation request associated to the request_uri when applicable.
+
+Files/Classes/methods affected:
+
+* org.keycloak.protocol.oidc.endpoints.request.AuthorizationEndpointRequestParserProcessor
```

Comment:
Current keycloak supports Request Object and its parameter `request_uri` the same as PAR. Therefore, keycloak needs to distinguish whether the authorization request with `request_uri` is Request Object or PAR.